Improve project menu

MikhailArkhipov opened the pull request:
Includes fix for #829

MikhailArkhipov commented:
![image](https://cloud.githubusercontent.com/assets/12820357/13866910/296b3ba8-ec77-11e5-8c77-13cb7825e9fc.png)

smortaz commented:
i didnt quite follow the thread - do individual files in a project get:

- Open in File Explorer
- Copy path

?

they really should (or get equivalent)


On Thu, Mar 17, 2016 at 7:32 PM, Mikhail Arkhipov <notifications@github.com>
wrote:

> Includes fix for #829 <https://github.com/Microsoft/RTVS/issues/829>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/Microsoft/RTVS/pull/1293
> Commit Summary
>
>    - Project menu commands
>    - Project menu
>    - Improve project menu
>    - Convert item path to friendly
>    - Add items to project folders
>    - Merge branch 'master' of https://github.com/Microsoft/RTVS
>    - Pass workflow for tests
>    - CR feedback
>    - Also fix #829
>
> File Changes
>
>    - *M* src/Package/Impl/Commands/R/RPackageCommandId.cs
>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-0> (7)
>    - *M* src/Package/Impl/Images/RFileNode.png
>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-1> (0)
>    - *M* src/Package/Impl/Microsoft.VisualStudio.R.Package.csproj
>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-2> (16)
>    - *M* src/Package/Impl/Package.vsct
>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-3> (158)
>    - *A* src/Package/Impl/ProjectSystem/Commands/AddItemCommand.cs
>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-4> (57)
>    - *A* src/Package/Impl/ProjectSystem/Commands/AddRMarkdownCommand.cs
>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-5> (20)
>    - *A* src/Package/Impl/ProjectSystem/Commands/AddRScriptCommand.cs
>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-6> (20)
>    - *A* src/Package/Impl/ProjectSystem/Commands/CommandUtilities.cs
>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-7> (53)
>    - *A* src/Package/Impl/ProjectSystem/Commands/CopyItemPathCommand.cs
>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-8> (42)
>    - *R*
>    src/Package/Impl/ProjectSystem/Commands/ExcludeFromProjectCommand.cs
>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-9> (2)
>    - *A*
>    src/Package/Impl/ProjectSystem/Commands/OpenContainingFolderCommand.cs
>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-10> (38)
>    - *A*
>    src/Package/Impl/ProjectSystem/Commands/SetDirectoryHereCommand.cs
>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-11> (43)
>    - *A* src/Package/Impl/ProjectSystem/Commands/SourceFilesCommand.cs
>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-12> (41)
>    - *A* src/Package/Impl/Repl/Commands/SourceFileHelper.cs
>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-13> (65)
>    - *M* src/Package/Impl/Repl/Commands/SourceRScriptCommand.cs
>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-14> (31)
>    - *M* src/Package/Impl/Repl/Commands/WorkingDirectoryCommand.cs
>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-15> (21)
>    - *M* src/Package/Impl/Utilities/ProjectUtilities.cs
>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-16> (100)
>    - *A* src/Package/Impl/Utilities/SessionUtilities.cs
>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-17> (78)
>    - *A* src/Package/Impl/VSAppShell/Current.cs
>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-18> (4)
>    - *M* src/Package/Test/DataInspect/EvaluationWrapperTest.cs
>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-19> (2)
>    - *M* src/Package/Test/Repl/CurrentDirectoryTest.cs
>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-20> (5)
>    - *M* src/Package/TestApp/Data/VariableExplorerTest.cs
>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-21> (2)
>    - *M* src/R/Core/Test/Files/Parser/frametools.r.tree
>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-22> (1102)
>
> Patch Links:
>
>    - https://github.com/Microsoft/RTVS/pull/1293.patch
>    - https://github.com/Microsoft/RTVS/pull/1293.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/Microsoft/RTVS/pull/1293>
>

smortaz commented:
nice!

On Thu, Mar 17, 2016 at 7:33 PM, Sean Mortazavi <smortaz@gmail.com> wrote:

> i didnt quite follow the thread - do individual files in a project get:
>
> - Open in File Explorer
> - Copy path
>
> ?
>
> they really should (or get equivalent)
>
>
> On Thu, Mar 17, 2016 at 7:32 PM, Mikhail Arkhipov <
> notifications@github.com> wrote:
>
>> Includes fix for #829 <https://github.com/Microsoft/RTVS/issues/829>
>> ------------------------------
>> You can view, comment on, or merge this pull request online at:
>>
>>   https://github.com/Microsoft/RTVS/pull/1293
>> Commit Summary
>>
>>    - Project menu commands
>>    - Project menu
>>    - Improve project menu
>>    - Convert item path to friendly
>>    - Add items to project folders
>>    - Merge branch 'master' of https://github.com/Microsoft/RTVS
>>    - Pass workflow for tests
>>    - CR feedback
>>    - Also fix #829
>>
>> File Changes
>>
>>    - *M* src/Package/Impl/Commands/R/RPackageCommandId.cs
>>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-0> (7)
>>    - *M* src/Package/Impl/Images/RFileNode.png
>>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-1> (0)
>>    - *M* src/Package/Impl/Microsoft.VisualStudio.R.Package.csproj
>>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-2> (16)
>>    - *M* src/Package/Impl/Package.vsct
>>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-3> (158)
>>    - *A* src/Package/Impl/ProjectSystem/Commands/AddItemCommand.cs
>>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-4> (57)
>>    - *A* src/Package/Impl/ProjectSystem/Commands/AddRMarkdownCommand.cs
>>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-5> (20)
>>    - *A* src/Package/Impl/ProjectSystem/Commands/AddRScriptCommand.cs
>>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-6> (20)
>>    - *A* src/Package/Impl/ProjectSystem/Commands/CommandUtilities.cs
>>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-7> (53)
>>    - *A* src/Package/Impl/ProjectSystem/Commands/CopyItemPathCommand.cs
>>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-8> (42)
>>    - *R*
>>    src/Package/Impl/ProjectSystem/Commands/ExcludeFromProjectCommand.cs
>>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-9> (2)
>>    - *A*
>>    src/Package/Impl/ProjectSystem/Commands/OpenContainingFolderCommand.cs
>>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-10> (38)
>>    - *A*
>>    src/Package/Impl/ProjectSystem/Commands/SetDirectoryHereCommand.cs
>>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-11> (43)
>>    - *A* src/Package/Impl/ProjectSystem/Commands/SourceFilesCommand.cs
>>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-12> (41)
>>    - *A* src/Package/Impl/Repl/Commands/SourceFileHelper.cs
>>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-13> (65)
>>    - *M* src/Package/Impl/Repl/Commands/SourceRScriptCommand.cs
>>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-14> (31)
>>    - *M* src/Package/Impl/Repl/Commands/WorkingDirectoryCommand.cs
>>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-15> (21)
>>    - *M* src/Package/Impl/Utilities/ProjectUtilities.cs
>>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-16> (100)
>>    - *A* src/Package/Impl/Utilities/SessionUtilities.cs
>>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-17> (78)
>>    - *A* src/Package/Impl/VSAppShell/Current.cs
>>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-18> (4)
>>    - *M* src/Package/Test/DataInspect/EvaluationWrapperTest.cs
>>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-19> (2)
>>    - *M* src/Package/Test/Repl/CurrentDirectoryTest.cs
>>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-20> (5)
>>    - *M* src/Package/TestApp/Data/VariableExplorerTest.cs
>>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-21> (2)
>>    - *M* src/R/Core/Test/Files/Parser/frametools.r.tree
>>    <https://github.com/Microsoft/RTVS/pull/1293/files#diff-22> (1102)
>>
>> Patch Links:
>>
>>    - https://github.com/Microsoft/RTVS/pull/1293.patch
>>    - https://github.com/Microsoft/RTVS/pull/1293.diff
>>
>> —
>> You are receiving this because you are subscribed to this thread.
>> Reply to this email directly or view it on GitHub
>> <https://github.com/Microsoft/RTVS/pull/1293>
>>
>
>

MikhailArkhipov commented:
Multiple selection

![image](https://cloud.githubusercontent.com/assets/12820357/13866936/7338fbee-ec77-11e5-8dc9-5e847e0ccda4.png)

MikhailArkhipov commented:
I can hide Add... for multiple selection but not sure

AlexanderSher commented on src/Package/Impl/Utilities/ProjectUtilities.cs:
```diff
@@ -42,5 +52,95 @@ public static class ProjectUtilities {
             return result;
         }
+
+        public static EnvDTE.Project GetActiveProject() {
```
Please pass `VsAppShell.Current` as a parameter

AlexanderSher replied:
Actually, maybe `DTE` is better

MikhailArkhipov replied:
DTE is hard to mock though.

int19h replied:
PTVS has a bunch of DTE mocks that might be of use.

https://github.com/Microsoft/PTVS/tree/0a3d9d17008e8d9b8fe2cb40c6c501f13861ccdb/Common/Tests/MockVsTests

MikhailArkhipov replied:
Since it uses GetGlobalService, it is easily replaceable in tests by mock. Doesn't have to be passed, see
```
        private TestVsAppShell() {
            CompositionService = VsTestCompositionCatalog.Current.CompositionService;
            ExportProvider = VsTestCompositionCatalog.Current.ExportProvider;
            MainThread = UIThreadHelper.Instance.Thread;
            _sp = new TestServiceProvider();
        }
```

so basically just add DTE mock to the TestServiceProvider. Currently it has

```
        public TestServiceProvider() {
            _services.Add(typeof(SVsRegisterProjectTypes), new VsRegisterProjectGeneratorsMock());
            _services.Add(typeof(SVsRegisterEditors), VsRegisterEditorsMock.Create());
            _services.Add(typeof(IMenuCommandService), new MenuCommandServiceMock());
            _services.Add(typeof(SComponentModel), new ComponentModelMock(VsTestCompositionCatalog.Current));
            _services.Add(typeof(SVsTextManager), new TextManagerMock());
            _services.Add(typeof(SVsImageService), VsImageServiceMock.Create());
            _services.Add(typeof(SVsUIShell), new VsUiShellMock());
            _services.Add(typeof(SOleComponentManager), OleComponentManagerMock.Create());
            _services.Add(typeof(SVsSettingsManager), VsSettingsManagerMock.Create());
        }
```

AlexanderSher commented on src/Package/Impl/Utilities/ProjectUtilities.cs:
```diff
@@ -42,5 +52,95 @@ public static class ProjectUtilities {
+                    // Locate folder in the project
+                    var projectFolder = Path.GetDirectoryName(project.FullName);
+                    if (destinationPath.StartsWith(projectFolder, StringComparison.OrdinalIgnoreCase)) {
+                        ProjectItems projectItems = project.ProjectItems;
```
Don't change ProjectItems explicitly - file system should handle the change automatically

MikhailArkhipov replied:
this is not changing, just drilling into folder. DTE.ProjectItems is root, then pick folder from them, get ProjectItems (aka children) from it and so on until the destination is reached.

int19h commented on src/Package/Impl/Utilities/ProjectUtilities.cs:
```diff
@@ -42,5 +52,95 @@ public static class ProjectUtilities {
+            return null;
+        }
+
+        public static void AddNewItem(string templateName, string name, string extension, string destinationPath) {
```
Do we really have to do the whole unzip dance ourselves? I hoped VS would have provided a programmatic way to do the same thing that "Add New Item" does when user clicks OK...

int19h replied:
Actually, looking at what C# and C++ do, it doesn't just add the item right away when you click "Add Class". Instead, it opens the usual "Add New Item" dialog (which lets the user provide the name for the class), but pre-selects the class template. I assume other "Add ..." commands work in a similar fashion.

Both C# and C++ use `IVsAddProjectItemDlg.AddProjectItemDlg` for this purpose - there is a parameter there to pre-select a template.

MikhailArkhipov replied:
DTE only support unpacked templates. I could go with the dialog, but then it is as many clicks as regular Add Item. With file system it is simpler to add file and then rename if needed.

int19h replied:
But then we're not consistent with the rest of VS.

MikhailArkhipov replied:
No all things in VS have to be the same. Besides, we only have two items while C# has like 10.

int19h replied:
If it's labeled the same, there should be a very good reason for it to not behave the same. In this case, I simply don't see such a reason, especially given the amount of custom code that this unusual behavior requires compared to the regular behavior. And click-wise, it is not only not better, it's actually worse. For example, consider C# again. When I do Add Class, the sequence is as follows:

1. Right-click folder/project -> "Add Class"
2. Type the name of the class (focus will already be in the right field in Add New Item dialog).
3. Press Enter.

With the flow implemented here:

1. Right-click project/folder -> "Add Script"
2. Right-click on the newly added file again -> "Rename" (or click-click it slowly).
3. Type the name of the file.
4. Press Enter.

So basically this workflow _adds_ an extra two clicks in all scenarios, unless I'm okay with "rscript1.R" etc. Which is pretty much never the case outside of hello world type apps.

MikhailArkhipov replied:
PM call. I personally don't like C# behavior, I'd rather add and rename.

MikhailArkhipov replied:
Besides, R more 'scratch' type language (i.e. use file to hold some code for some time) that others so naming files in many cases may not be that important. RS does not show anything, just adds the file.

int19h replied:
@jflam, your call.

MikhailArkhipov replied:
John said it is OK to be different.

AlexanderSher commented on src/Package/Impl/Utilities/SessionUtilities.cs:
```diff
@@ -0,0 +1,78 @@
+
+namespace Microsoft.VisualStudio.R.Package.Utilities {
+    public static class SessionUtilities {
+        public static IRSession GetInteractiveSession() {
```
We should not create extensions like this one, cause it couples `Microsoft.R.Components` object to the `VsAppShell.Current` instance. That is bad because:
- Any instance of the object from `Microsoft.R.Components` can live longer than `VsAppShell` and created before it. Calling to `VsAppShell.Current` before it is initialized in Package breaks pretty much everything. 
- `VsAppShell.Current` implicitly adds dependency to the IVs* to the code that doesn't really need it (we have dozens of commands in `Package` that can be moved to the `Microsoft.R.Components`).
- Any code that uses `VsAppShell.Current` requires isolated non-parallel run, lots of fakes and VsRHostScript even if the test has no UI involved.

AlexanderSher commented on src/Package/Impl/Utilities/SessionUtilities.cs:
```diff
@@ -0,0 +1,78 @@
+            return interactiveWorkflow.RSession;
+        }
+
+        public static async Task<string> GetRWorkingDirectoryAsync(IRInteractiveWorkflow workflow = null) {
```
This code can be moved to the Microsoft.R.Host.Client:
```{csharp}
public static async Task<string> GetRWorkingDirectoryAsync(this IRSession session) {
    if (session.IsHostRunning) {
        await TaskUtilities.SwitchToBackgroundThread();
        try {
            using (var evaluation = await session.BeginEvaluationAsync(false)) {
               return await evaluation.GetWorkingDirectory();
            }
        } catch (OperationCanceledException) { }
    }
    return null;
}
```

Also, `IRInteractiveWorkflow` is usually assumed to be not null.

AlexanderSher commented on src/Package/Impl/Utilities/SessionUtilities.cs:
```diff
@@ -0,0 +1,78 @@
+            return null;
+        }
+
+        public static string GetRShortenedPathName(string name, string userDirectory) {
```
This belongs to `Microsoft.R.Host.Client.RStringExtensions` I think.

MikhailArkhipov replied:
I'll make it SessionExtensions

AlexanderSher commented on src/Package/Impl/Utilities/SessionUtilities.cs:
```diff
@@ -0,0 +1,78 @@
+            return null;
+        }
+
+        public static async Task<string> GetRUserDirectoryAsync(IRInteractiveWorkflow workflow = null) {
```
Same with this one

AlexanderSher commented on src/Package/Impl/VSAppShell/Current.cs:
```diff
@@ -0,0 +1,4 @@
+﻿namespace VSAppShell {
+    internal class Current {
```
What is this?

MikhailArkhipov replied:
Whoops. misclick. Instead of 'add using' apparently clicked 'generate class'

AlexanderSher commented on src/Package/Impl/ProjectSystem/Commands/CommandUtilities.cs:
```diff
@@ -0,0 +1,53 @@
+using Microsoft.VisualStudio.ProjectSystem.Designers;
+
+namespace Microsoft.VisualStudio.R.Package.ProjectSystem.Commands {
+    internal static class CommandUtilities {
```
I don't like this name. None of these methods is about commands.

MikhailArkhipov replied:
ProjectTreeExtensions and move to the project system assembly?

AlexanderSher replied:
Yeah, great idea!

AlexanderSher commented on src/Package/Impl/ProjectSystem/Commands/SetDirectoryHereCommand.cs:
```diff
@@ -0,0 +1,43 @@
+                if (!string.IsNullOrEmpty(path)) {
+                    var o = new object();
+
+                    var interactiveWorkflowProvider = VsAppShell.Current.ExportProvider.GetExportedValue<IRInteractiveWorkflowProvider>();
```
`SetDirectoryHereCommand` is exported through MEF, so `IRInteractiveWorkflowProvider` this can be obtained in ctor

AlexanderSher commented on src/Package/Impl/ProjectSystem/Commands/SetDirectoryHereCommand.cs:
```diff
@@ -0,0 +1,43 @@
+                    var interactiveWorkflow = interactiveWorkflowProvider.GetOrCreate();
+                    var controller = ReplCommandController.FromTextView(interactiveWorkflow.ActiveWindow.TextView);
+
+                    controller.Invoke(RGuidList.RCmdSetGuid, RPackageCommandId.icmdSetWorkingDirectory, path, ref o);
```
Not a part of this change, but it would be great to wire up these `ICommandGroupHandler` and `IAsyncCommandGroupHandler` interfaces to our controller-based invocation in general.

AlexanderSher commented on src/Package/Impl/ProjectSystem/Commands/SourceFilesCommand.cs:
```diff
@@ -0,0 +1,41 @@
+        public bool TryHandleCommand(IImmutableSet<IProjectTree> nodes, long commandId, bool focused, long commandExecuteOptions, IntPtr variantArgIn, IntPtr variantArgOut) {
+            if (commandId == RPackageCommandId.icmdSourceSelectedFiles) {
+                var rFiles = nodes.GetSelectedNodesPaths().Where(x =>
+                               Path.GetExtension(x).Equals(".r", StringComparison.OrdinalIgnoreCase) &&
```
we have `Microsoft.Common.Core.StringExtensions.EqualsIgnoreCase`.

AlexanderSher commented on src/Package/Impl/Repl/Commands/SourceFileHelper.cs:
```diff
@@ -0,0 +1,65 @@
+        }
+
+        private static bool IsDebugging() {
+            var debugger = VsAppShell.Current.GetGlobalService<IVsDebugger>(typeof(SVsShellDebugger));
```
We have IDebuggerModeTracker. We can add another property to it.

AlexanderSher replied:
Btw, @int19h, can't we get this information from `IRSession` or `DebugSession` directly?

int19h replied:
No, because `DebugSession` is always there (its lifetime doesn't correspond to VS debugging or not debugging, since it's also used by VE etc).

int19h replied:
Although I guess the real question that this code should ask is, "are breakpoints enabled", in which case `DebugSession` can track that (it doesn't currently - it just passes it on to R code which does the actual tracking - but it would be easy to add).

AlexanderSher commented on src/Package/Impl/Repl/Commands/SourceFileHelper.cs:
```diff
@@ -0,0 +1,65 @@
+namespace Microsoft.VisualStudio.R.Package.Repl.Commands {
+    internal static class SourceFileHelper {
+        public static void SourceFiles(IEnumerable<string> files) {
+            VsAppShell.Current.AssertIsOnMainThread();
```
Why do you need a main thread here?

MikhailArkhipov replied:
Enqueue/Execute Expression use nteractiveWindow and nteractiveWindow.TextView that are not thread safe and are UI bound.

AlexanderSher replied:
But you access Enqueue/ExecuteExpression inside `DispatchOnUIThread`

MikhailArkhipov replied:
IsDebugging calls ```GetGlobalService<IVsDebugger>```

AlexanderSher commented on src/Package/Impl/Repl/Commands/SourceFileHelper.cs:
```diff
@@ -0,0 +1,65 @@
+
+namespace Microsoft.VisualStudio.R.Package.Repl.Commands {
+    internal static class SourceFileHelper {
+        public static void SourceFiles(IEnumerable<string> files) {
```
I think we can add this method to the `IRInteractiveWorkflowOperations` interface

AlexanderSher replied:
And make it async

AlexanderSher commented on src/Package/Impl/Repl/Commands/SourceFileHelper.cs:
```diff
@@ -0,0 +1,65 @@
+
+            Task.Run(async () => {
+                foreach (var f in files) {
+                    list.Add(await SessionUtilities.GetRShortenedPathNameAsync(f));
```
This is actually  a good example why we have separated `IRSession.BeginEvaluationAsync` into separate method. This code schedules `files.Count()` tasks into `RSession` evaluation queue, without any guarantee that they will be executed one after another.
Instead, this code should call `IRSession.BeginEvaluationAsync` once for the whole loop.

AlexanderSher replied:
Not to mention that `GetRShortenedPathNameAsync` schedules 2 tasks itself - in `GetRUserDirectoryAsync` and in `GetRShortenedPathName`

int19h replied:
Is it a correctness issue here, or a perf issue?

Anyway, it sounds like you're suggesting that any helper method that performs evals should have an overload that accepts `IRSessionEvaluation`, to reuse an existing one?

MikhailArkhipov replied:
They don't have to be in order

AlexanderSher replied:
User can change his current directory while this loop is running, so part of the paths will be pointing to the wrong place. And we actually don't need to call `GetRUserDirectoryAsync` on every step of the loop.

As for the helper methods, I think that for this type of extensions we should have them for `IRSessionEvaluation` rather than for `IRSession`.

MikhailArkhipov replied:
Theoretically yes. Practically, it is probably 3-4 files and it is unlikely user will be clicking around while files are getting sourced.

MikhailArkhipov replied:
I'll make batch method extension

AlexanderSher commented on src/Host/Client/Impl/Extensions/SessionExtensions.cs:
```diff
@@ -0,0 +1,35 @@
+using Microsoft.R.Host.Client.Session;
+
+namespace Microsoft.R.Host.Client {
+    public static class SessionExtensions {
```
We already have class called `RSessionExtensions`
It is in the `Microsoft.R.Host.Client.Session`. We probably should merge them.

AlexanderSher commented:
Sorry guys, I'm still not feeling well, so I'm done for today.

MikhailArkhipov commented:
Addressed most of the feedback, added Open Command Prompt. If we are settled on general interfaces and code locations, I will be working on some tests for the new commands.

MikhailArkhipov commented:
Ping... We can change Add New Item story based on user feedback.

AlexanderSher commented on src/Host/Client/Impl/Session/RSessionExtensions.cs:
```diff
@@ -23,5 +24,57 @@ public static class RSessionExtensions {
+
+        public static async Task<IEnumerable<string>> GetRShortenedPathNamesAsync(this IRSession session, IEnumerable<string> names) {
+            var userDirectory = await session.GetRUserDirectoryAsync();
+            var list = new List<string>();
```
```{csharp}
return names.Select(n => GetRShortenedPathName(n, userDirectory));
```

AlexanderSher commented on src/Host/Client/Impl/Session/RSessionExtensions.cs:
```diff
@@ -23,5 +24,57 @@ public static class RSessionExtensions {
+            return GetRShortenedPathName(name, userDirectory);
+        }
+
+        public static async Task<IEnumerable<string>> GetRShortenedPathNamesAsync(this IRSession session, IEnumerable<string> names) {
```
I don't like the name of the method. Maybe something like `MakeRelativeToUserDirectory`?

MikhailArkhipov replied:
OK

AlexanderSher commented:
:+1:
